src/pydap/model.py

```diff
@@ -307,55 +392,100 @@ def __getattr__(self, attr):
+        return (key in self.visible_keys)
+
+    # __iter__, __getitem__, __len__ are required for Mapping
+    # From these, keys, items, values, get, __eq__,
+    # and __ne__ are obtained.
     def __iter__(self):
-        for key in self._keys:
-            x = self._dict[key]
-            if isinstance(x, binary_type):
-                yield x.tostring().decode('utf-8')
-            else:
-                yield x
+        for key in self._dict.keys():
+            if key in self.visible_keys:
+                yield key
 
-    children = __iter__
+    def all_keys(self):
+        return iter(self._dict.keys())
 
-    def __setitem__(self, key, item):
-        key = quote(key)
-        if key != item.name:
-            raise KeyError(
-                'Key "%s" is different from variable name "%s"!' %
-                (key, item.name))
+    def all_items(self):
+        return iter(self._dict.items())
 
-        if key in self._keys:
-            self._keys.pop(self._keys.index(key))
-        self._keys.append(key)
-        self._dict[key] = item
+    def all_values(self):
+        return iter(self._dict.values())
 
-        # Set item id.
-        item.id = '%s.%s' % (self.id, item.name)
+    def _set_visible_keys(self, keys):
+        self._visible_keys = keys
 
-    def __getitem__(self, key):
+    def _get_visible_keys(self):
+        return self._visible_keys
+    visible_keys = property(_get_visible_keys, _set_visible_keys)
+
+    def _getitem_string(self, key):
+        """ Assume that key is a string type """
         try:
-            return self._dict[quote(key)]
+            x = self._dict[quote(key)]
+            if isinstance(x, binary_type):
+                x = x.tostring().decode('utf-8')
+            return x
         except KeyError:
             splitted = key.split('.')
             if len(splitted) > 1:
                 try:
-                    return (self
-                            .__getitem__(splitted[0])['.'.join(splitted[1:])])
+                    return self[splitted[0]]['.'.join(splitted[1:])]
                 except KeyError:
-                    return self.__getitem__('.'.join(splitted[1:]))
+                    return self['.'.join(splitted[1:])]
             else:
                 raise
 
-    def __delitem__(self, key):
-        self._dict.__delitem__(key)
-        self._keys.remove(key)
+    def _getitem_string_tuple(self, key):
+        """ Assume that key is a tuple of strings """
+        out = type(self)(self.name, data=self.data,
+                         attributes=self.attributes.copy())
+        for name in key:
+            out[name] = copy.copy(self._getitem_string(name))
+        return out
 
-    def keys(self):
-        """Method to emulate a dictionary, returning keys."""
-        return self._keys[:]
+    def __getitem__(self, key):
+        if isinstance(key, string_types):
+            return self._getitem_string(key)
+        elif (isinstance(key, tuple) and
```

Comment:
It might be a better idea to use `list` instead of `tuple` for this special type. Here's why: Python translates `x[y, z]` into `x.__getitem__((y, z))`, automatically turning multiple arguments in `[]` into a tuple. So it would be harder to use `[]` accidentally.

That said, let's save this one for later, given that apparently this was supported in earlier versions of pydap.